### PR TITLE
Purecap kernel configuration and build system changes

### PR DESCRIPTION
**Pull request by qwattash**
I reworked the history and grouped various changes to configuration and makefiles/options here.

### CODE REVIEW

File: sys/conf/Makefile.mips [hunk 1]
```diff
@@ -62,6 +64,8 @@ TRAMP_ELFSIZE=64
 .else
 TRAMP_ELFSIZE=32
 .endif
+# XXX-AM: Hack to clear trampoline flags
+TRAMP_EXTRA_LDFLAGS=${TRAMP_LDFLAGS:N-fuse-ld*}
```
Comment (jrtc27) [inline sys/conf/Makefile.mips:68]: What's this for?

File: sys/conf/kern.mk [hunk 2]
```diff
@@ -269,13 +269,21 @@ CFLAGS+=	-gdwarf-2
 # CHERI purecap kernel flags
 #
 .if ${MACHINE_ABI:Mpurecap}
+CWARNFLAGS+=	-Werror=implicit-function-declaration
 
+.ifdef CHERI_USE_CAP_TABLE
```
Comment (jrtc27) [inline sys/conf/kern.mk:274]:
```suggestion
.if defined(CHERI_USE_CAP_TABLE)
```

File: sys/mips/conf/BERI_DE4_2CORE_MDROOT [hunk 3]
```diff
@@ -8,6 +8,7 @@
 
 include "BERI_DE4_MDROOT"
 
+machine		mips mips64
```
Comment (jrtc27) [inline sys/mips/conf/BERI_DE4_2CORE_MDROOT:11]: Does this not belong in a BASE/std. file?

File: sys/mips/conf/CHERI_PURECAP_DE4_MFS_ROOT_BASE [hunk 4]
```diff
@@ -0,0 +1,38 @@
+#
```
Comment (jrtc27) [inline sys/mips/conf/CHERI_PURECAP_DE4_MFS_ROOT_BASE:1]: These should all be very boring files that inherit from a base CHERI config. The BERI configs are already a mess, we don't need to add another whole set of configs that duplicate all the options.

File: sys/mips/conf/std.CHERI [hunk 5]
```diff
@@ -5,10 +5,11 @@
 # $FreeBSD$
 #
 nomakeoptions	ARCH_FLAGS
-makeoptions	ARCH_FLAGS="-target mips64-unknown-freebsd -march=beri -mabi=64 -mcpu=beri -cheri -Wno-unused-command-line-argument"
+makeoptions	ARCH_FLAGS="-march=beri -mcpu=beri"
 options 	CPU_CHERI
 nooptions	COMPAT_FREEBSD32
 options 	COMPAT_FREEBSD64
+options 	BOOTVERBOSE=1
```
Comment (jrtc27) [inline sys/mips/conf/std.CHERI:12]: Behavioural change?

File: sys/mips/conf/std.CHERI_PURECAP [hunk 6]
```diff
@@ -0,0 +1,50 @@
...
+
+include		"std.CHERI"
+
+machine		mips mips64c128
+
+nomakeoptions 	ARCH_FLAGS
+makeoptions 	ARCH_PIC_FLAGS="-fpic"
+makeoptions 	ARCH_FLAGS="-mcpu=cheri128 -mxcaptable"
+makeoptions 	CFLAGS+="-gdwarf-5"
+makeoptions 	WERROR="-Wno-error"
```
Comment (jrtc27) [inline sys/mips/conf/std.CHERI_PURECAP:17]: Should this not be in the .mk files?

File: sys/mips/conf/std.CHERI_PURECAP [hunk 7]
```diff
@@ -0,0 +1,50 @@
...
+makeoptions 	KERNLOADADDR=0xffffffff80100000
+makeoptions 	CHERI_USE_CAP_TABLE=plt
+makeoptions	CHERI_SUBOBJECT_BOUNDS=subobject-safe
+makeoptions   LDSCRIPT_NAME=ldscript.mips.cheri
```
Comment (jrtc27) [inline sys/mips/conf/std.CHERI_PURECAP:21]: Ditto

File: sys/mips/conf/std.CHERI_PURECAP [hunk 8]
```diff
@@ -0,0 +1,50 @@
...
+
+# purecap kernel options
+options		CHERI_PURECAP_KERNEL
```
Comment (jrtc27) [inline sys/mips/conf/std.CHERI_PURECAP:24]: Formatting is a mess from here onwards

File: sys/mips/conf/std.CHERI_PURECAP [hunk 9]
```diff
@@ -0,0 +1,50 @@
...
+# filesystems
+options 	TMPFS
+options		FFS
+options		CD9660
+options		MSDOSFS
+nooptions		NFSD, NFSCL
+
+options		HWPMC_HOOKS
+device		hwpmc
+device		hwpmc_beri
```
Comment (jrtc27) [inline sys/mips/conf/std.CHERI_PURECAP:50]: All of this (except the nooptions) belongs in a base MIPS/BERI/CHERI config, it's not purecap-specific.

File: sys/riscv/conf/CHERI-PURECAP-FETT [hunk 10]
```diff
@@ -0,0 +1,9 @@
+#NO_UNIVERSE
+
+include "FETT"
+include "std.CHERI-PURECAP"
+
+options 	ENABLE_PAST_LOCAL_VULNERABILITIES
+options 	ENABLE_PAST_REMOTE_VULNERABILITIES
```
Comment (jrtc27) [inline sys/riscv/conf/CHERI-PURECAP-FETT:7]: These don't make sense to commit without the corresponding vulnerabilities and config options available.

File: sys/riscv/conf/CHERI-PURECAP-QEMU [hunk 11]
```diff
@@ -0,0 +1,9 @@
+#NO_UNIVERSE
+
+include "QEMU"
+include "std.CHERI-PURECAP"
+
+options 	ENABLE_PAST_LOCAL_VULNERABILITIES
+options 	ENABLE_PAST_REMOTE_VULNERABILITIES
```
Comment (jrtc27) [inline sys/riscv/conf/CHERI-PURECAP-QEMU:7]: We probably don't actually want this in the normal QEMU kernel.

File: sys/riscv/conf/std.CHERI-PURECAP [hunk 12]
```diff
@@ -0,0 +1,8 @@
+include "std.CHERI"
+
+machine		riscv	riscv64c
+
+makeoptions	NO_MODULES=yes
```
Comment (jrtc27) [inline sys/riscv/conf/std.CHERI-PURECAP:5]: Hmm, this feels more like a thing for .mk when `${MACHINE_ABI:Mpurecap}` in general, not an arch-specific kernel config option.

File: sys/mips/conf/std.CHERI (hunk 5 above)
Comment (qwattash) [reply sys/mips/conf/std.CHERI:12]: I missed that, I think I added it at some point because the purecap kernel did not produce `start_init` output, causing false timeouts in jenkins.

File: sys/mips/conf/std.CHERI_PURECAP (hunk 6 above)
Comment (qwattash) [reply sys/mips/conf/std.CHERI_PURECAP:17]: I moved most of those. I left WERROR as we would ideally remove it

File: sys/mips/conf/std.CHERI_PURECAP (hunk 7 above)
Comment (qwattash) [reply sys/mips/conf/std.CHERI_PURECAP:21]: I'm tempted to keep these to have a more explicit configuration for those parameters, albeit a bit more verbose. I can move the linker script though.

File: sys/conf/Makefile.mips (hunk 1 above)
Comment (qwattash) [reply sys/conf/Makefile.mips:68]: It is an old hack that should have been reverted. Removed now.

File: sys/conf/kern.mk [hunk 13]
```diff
@@ -3,7 +3,7 @@
 #
 # Warning flags for compiling the kernel and components of the kernel:
 #
-CWARNFLAGS?=	-Wall -Wredundant-decls -Wnested-externs -Wstrict-prototypes \
+CWARNFLAGS=	-Wall -Wredundant-decls -Wnested-externs -Wstrict-prototypes \
```
Comment (brooksdavis) [inline sys/conf/kern.mk:6]: Is this needed?

Comment (qwattash) [reply sys/conf/kern.mk:6]: I think so, this was added in bb8961f6f257c2ce1c4 and I think bsd.cpu.mk is still changing the CWARNFLAGS when in purecap.

Comment (brooksdavis) [reply sys/conf/kern.mk:6]: Ok, this probably isn't upstreamable, but it's likely fine for now.

**Review by brooksdavis — APPROVED** [review]

File: sys/riscv/conf/std.CHERI-PURECAP (hunk 12 above)
Comment (bsdjhb) [reply sys/riscv/conf/std.CHERI-PURECAP]: Please no, it's a lot easier to work with in my branch where I'm working on kldload if it stays this way for now.

Comment (jrtc27) [reply sys/riscv/conf/std.CHERI-PURECAP]: Well, no, it doesn't belong in the kernel config, and if you use `?=` in a .mk file then you can continue to set it on the command line whilst only having a single copy of the variable definition shared by all architectures in the "right" place.

Comment (bsdjhb) [reply sys/riscv/conf/std.CHERI-PURECAP]: You can't unset variables on the command line.  And the value doesn't matter, just if it is defined or not (see kern.*.mk where all tests are just `if !defined(NO_MODULES)`.

Comment (jrtc27) [reply sys/riscv/conf/std.CHERI-PURECAP]: Ugh... I hate that pattern.

File: sys/mips/conf/CHERI_PURECAP_FUZZ_DE4_MFS_ROOT [hunk 14]
```diff
@@ -0,0 +1,16 @@
...
+# RISC Instructions) FPGA soft core, as configured in its Terasic DE-4
+# reference configuration - 128bit version.
+#
+# This configuration is the same as the pure capability kernel with MFS root,
+# with the addition of kernel coverage.
+#
+# $FreeBSD$
+#
+
+include		"CHERI128_PURECAP_DE4_MFS_ROOT"
```
Comment (brooksdavis) [inline sys/mips/conf/CHERI_PURECAP_FUZZ_DE4_MFS_ROOT:12]: missing ident (and in all the other FUZZ kernels)

File: sys/mips/conf/std.CHERI_PURECAP [hunk 15]
```diff
@@ -0,0 +1,34 @@
...
+makeoptions 	CHERI_USE_CAP_TABLE=plt
+makeoptions	CHERI_SUBOBJECT_BOUNDS=subobject-safe
+makeoptions	LDSCRIPT_NAME=ldscript.mips.cheri
+
+# purecap kernel options
+options		CHERI_PURECAP_KERNEL
+
+# debugging
+options		KDB_TRACE # stacktrace on panic
+nooptions	DDB
```
Comment (brooksdavis) [inline sys/mips/conf/std.CHERI_PURECAP:22]: Is this disabled because it doesn't work? We'll want to fix that soon.

File: sys/mips/conf/CHERI_PURECAP_FUZZ_DE4_MFS_ROOT (hunk 14 above)
Comment (brooksdavis) [reply sys/mips/conf/CHERI_PURECAP_FUZZ_DE4_MFS_ROOT:12]: They should probably also have `#NO_UNIVERSE` tags

File: sys/mips/conf/std.CHERI_PURECAP (hunk 15 above)
Comment (bsdjhb) [reply sys/mips/conf/std.CHERI_PURECAP:22]: It works on RISC-V.  I have not looked at what it would take to fix it on MIPS, probably not too much further work.

Comment (qwattash) [reply sys/mips/conf/std.CHERI_PURECAP:22]: It has been untested for a long time, probably needs some work.

File: sys/mips/conf/BERI_NETFPGA_MDROOT [hunk 16]
```diff
@@ -8,6 +8,8 @@
 
 include "std.BERI"
 
+machine		mips mips64
```
Comment (arichardson) [inline sys/mips/conf/BERI_NETFPGA_MDROOT:11]: Can the machine option go into std.BERI?

File: sys/mips/conf/CHERI_PURECAP_DE4_MFS_ROOT_FUZZ [hunk 17]
```diff
@@ -0,0 +1,19 @@
...
+# reference configuration - 128bit version.
+#
+# $FreeBSD$
+#
+
+include		"CHERI_PURECAP_DE4_MFS_ROOT"
+ident		CHERI128_PURECAP_DE4_MFS_ROOT_FUZZ
+
+# Disable slow things
+include "std.BENCHMARK"
```
Comment (arichardson) [inline sys/mips/conf/CHERI_PURECAP_DE4_MFS_ROOT_FUZZ:14]: Do you also want a fuzzing kernel that has assertions enabled?

File: sys/conf/ldscript.mips.cheri [hunk 18]
```diff
@@ -0,0 +1,319 @@
...
+  {
+    KEEP (*crtbegin.o(.dtors))
+    KEEP (*(EXCLUDE_FILE (*crtend.o ) .dtors))
+    KEEP (*(SORT(.dtors.*)))
+    KEEP (*(.dtors))
+  }
+  .plt      : { *(.plt)	}
+  _gp = ALIGN(16) + 0x7ff0;
+  .got		  : { *(.got.plt) *(.got) }
+  _captable = ALIGN(0x20);
```
Comment (arichardson) [inline sys/conf/ldscript.mips.cheri:241]: LLD should already be defining `__cap_table_start` and `__cap_table_end`. Can those be used instead? There is also `_CHERI_CAPABILITY_TABLE_` but there's no end symbol for that.

Is there anything else that's different from the base linker script?
The
```
  __cap_relocs : {
    *(__cap_relocs)
  }
```
section should be fine to add to the MIPS linker script since it won't do anything if the section doesn't exist.

Comment (qwattash) [reply sys/conf/ldscript.mips.cheri:241]: The linker script is from a time when we did not have as much stability, probably can be ditched now. I'm finishing a merge and then go back to testing that.

File: sys/mips/conf/CHERI_PURECAP_DE4_MFS_ROOT_FUZZ (hunk 17 above)
Comment (qwattash) [reply sys/mips/conf/CHERI_PURECAP_DE4_MFS_ROOT_FUZZ:14]: The benchmark here was used mostly to have decent run times with kcov, in theory the fuzz kernel should probably only exist with assertions enabled

Comment (qwattash) [reply sys/mips/conf/CHERI_PURECAP_DE4_MFS_ROOT_FUZZ:14]: This was actually duplicating another variant. Removed it.

File: sys/conf/ldscript.mips.cheri (hunk 18 above)
Comment (qwattash) [reply sys/conf/ldscript.mips.cheri:241]: There are a bunch of extra ALIGN directives that have been added as a result of fb7f00b8c4aea68ceaf041f2340494b53fab6645
Some of the start/end symbols have been defined e5e258346f4edc1926223fef9f776fc723b39ea8 for capability relocations.
There is also some explicit padding that I believe was needed for the FPGA image ef3bc0a1cb41feec995739af7c4511cfda33fff5. If we are comfortable moving everything to the normal MIPS ldscript I can merge them. Otherwise I can just change the start/end section names I'm using.

Comment (arichardson) [reply sys/conf/ldscript.mips.cheri:241]: I'd be happy to have it merged to the normal MIPS ldscript. If people feel that this is not okay, I'd vote for adding a ldscript.mips.template with #ifdefs that we turn into the appropriate one with cpp.

**Review by brooksdavis — APPROVED** [review]

**Comment by brooksdavis** [comment]
A number of files lack line breaks before EOL (easy to spot in GitHub), but that's not a huge issue.

File: sys/conf/kern.pre.mk [hunk 19]
```diff
@@ -50,6 +50,10 @@ NM?=		nm
 OBJCOPY?=	objcopy
 SIZE?=		size
 
+.if ${MACHINE_ABI:Mpurecap}
```
Comment (bsdjhb) [inline sys/conf/kern.pre.mk:53]: I'd be happier to just leave this commit off of dev for now.  I will probably get all the fixes for modules merged to dev by the time RISC-V purecap kernels are built anyway, and this just adds more work to revert later if we keep it.

**Comment by jrtc27** [comment]
> A number of files lack line breaks before EOL (easy to spot in GitHub), but that's not a huge issue.

"Only" 3 of them, though it would be good to have those fixed so it's not in the history. After all, if a file doesn't end in a newline character then it's not strictly a text file according to POSIX... :)

File: sys/conf/Makefile.mips [hunk 20]
```diff
@@ -46,9 +50,14 @@ ARCH_FLAGS?=-march=mips64
 .else
 ARCH_FLAGS?=-march=mips32
 .endif
+
```
Comment (jrtc27) [inline sys/conf/Makefile.mips:53]: Unnecessary diff.

File: sys/conf/Makefile.mips [hunk 21]
```diff
@@ -46,9 +50,14 @@ ARCH_FLAGS?=-march=mips64
...
 ARCH_FLAGS+=-mabi=${MIPS_ABI}
-EXTRA_FLAGS=-fno-pic -mno-abicalls -G0 -DKERNLOADADDR=${KERNLOADADDR}
-EXTRA_FLAGS+=-${MIPS_ENDIAN}
+.if ${MACHINE_ABI:Mpurecap}
+EXTRA_FLAGS=-fpic -mcpu=cheri128 -mxcaptable
```
Comment (jrtc27) [inline sys/conf/Makefile.mips:56]: I am surprised we need `-mcpu=cheri128` given it's not there for a hybrid kernel. Modules also don't use this file (right?) so are clearly getting it from the right place. Is this just outdated?

**Comment by brooksdavis** [comment]
> > A number of files lack line breaks before EOL (easy to spot in GitHub), but that's not a huge issue.
> 
> "Only" 3 of them, though it would be good to have those fixed so it's not in the history. After all, if a file doesn't end in a newline character then it's not strictly a text file according to POSIX... :)

1, 2, many :)

File: sys/conf/Makefile.mips [hunk 22]
```diff
@@ -30,7 +30,11 @@ S=	../../..
 
 INCLUDES+= -I$S/contrib/libfdt
 
+.if ${MACHINE_ABI:Mpurecap}
+LDSCRIPT_NAME?=ldscript.$M.cheri
```
Comment (jrtc27) [inline sys/conf/Makefile.mips:34]: Isn't this wrong now?

File: sys/conf/ldscript.mips [hunk 23]
```diff
@@ -51,6 +51,10 @@ SECTIONS
   {
     KEEP (*(.fini))
   } =0x1000000
+  .fill1     :
```
Comment (jrtc27) [inline sys/conf/ldscript.mips:54]: What is this?

File: sys/conf/ldscript.mips [hunk 24]
```diff
@@ -176,6 +180,12 @@ SECTIONS
     }
   .rel.plt       : { *(.rel.plt)		}
   .rela.plt      : { *(.rela.plt)		}
+
+  PROVIDE (_rel_dyn = ALIGN(0x8));
```
Comment (jrtc27) [inline sys/conf/ldscript.mips:184]: Uh, no, you shouldn't need these. Look them up in `_DYNAMIC`, it'll give you size information too.

File: sys/conf/ldscript.mips [hunk 25]
```diff
@@ -176,6 +180,12 @@ SECTIONS
...
+  .rel.dyn	 : { *(.rel.dyn)		}
+  PROVIDE (_erel_dyn = .);
+  .rela.dyn	 : { *(.rela.dyn)		}
+  __cap_relocs : { *(__cap_relocs)		}
```
Comment (jrtc27) [inline sys/conf/ldscript.mips:188]: RISC-V didn't need this...

File: sys/conf/ldscript.mips [hunk 26]
```diff
@@ -223,6 +233,7 @@ SECTIONS
   .plt      : { *(.plt)	}
   _gp = ALIGN(16) + 0x7ff0;
   .got		  : { *(.got.plt) *(.got) }
+  .captable ALIGN(0x20): { *(.captable) }
```
Comment (jrtc27) [inline sys/conf/ldscript.mips:236]: Ditto...

File: sys/conf/ldscript.mips [hunk 27]
```diff
@@ -260,7 +271,7 @@ SECTIONS
       .bss section disappears because there are no input sections.  */
    . = ALIGN(64 / 8);
   }
-  . = ALIGN(64 / 8);
+  . = ALIGN(0x100000);
```
Comment (jrtc27) [inline sys/conf/ldscript.mips:274]: Do we want to be padding like this on hybrid and vanilla kernels?

File: sys/conf/ldscript.mips [hunk 28]
```diff
@@ -260,7 +271,7 @@ SECTIONS
       .bss section disappears because there are no input sections.  */
    . = ALIGN(64 / 8);
```
Comment (jrtc27) [inline sys/conf/ldscript.mips:272]: Should this not match below?

File: sys/conf/options [hunk 29]
```diff
@@ -1019,6 +1019,7 @@ LINDEBUGFS
 
 # CHERI
 CPU_CHERI	opt_global.h
+CHERI_PURECAP_KERNEL		opt_global.h
```
Comment (jrtc27) [inline sys/conf/options:1022]: Can we keep this option out of dev?

File: sys/conf/Makefile.mips (hunk 22 above)
Comment (jrtc27) [reply sys/conf/Makefile.mips:34]: (and the commit message is similarly outdated)

**Review by jrtc27 — CHANGES_REQUESTED** [review]

**Comment by jrtc27** [comment]
> > > A number of files lack line breaks before EOL (easy to spot in GitHub), but that's not a huge issue.
> > 
> > 
> > "Only" 3 of them, though it would be good to have those fixed so it's not in the history. After all, if a file doesn't end in a newline character then it's not strictly a text file according to POSIX... :)
> 
> 1, 2, many :)

I actually missed a 4th, so even more many

File: sys/conf/kern.pre.mk (hunk 19 above)
Comment (qwattash) [reply sys/conf/kern.pre.mk:53]: Dropped.

File: sys/conf/Makefile.mips (hunk 21 above)
Comment (qwattash) [reply sys/conf/Makefile.mips:56]: I think this comes from a very old configuration file. I tested the builds again, turns out that `-mxcaptable` is also not required anymore.

File: sys/conf/ldscript.mips (hunk 23 above)
Comment (qwattash) [reply sys/conf/ldscript.mips:54]: It was a workaround to pad the FPGA image for the DE4. The correct thing to do is to generate a binary image for berictl to load with `objcopy`, but we have not been doing so. Perhaps this should not be in dev and I should have cheribuild use objcopy. I remember there was a reason I did not do that in the first place, but can't entirely recall it now.

File: sys/conf/options (hunk 29 above)
Comment (qwattash) [reply sys/conf/options:1022]: Not sure, it is currently used to conditionally compile a bunch of files. We generally have most of the changes in common files but for MIPS I duplicated a bunch of assembly files because everything changed.

File: sys/conf/ldscript.mips (hunk 27 above)
Comment (qwattash) [reply sys/conf/ldscript.mips:274]: If it hurts image size too much I can pass the file through cpp, not sure if it is worth it for 1MiB padding at most.

File: sys/conf/ldscript.mips (hunk 24 above)
Comment (qwattash) [reply sys/conf/ldscript.mips:184]: I think these were from earlier linker versions. I have now updated it.

File: sys/conf/ldscript.mips (hunk 28 above)
Comment (qwattash) [reply sys/conf/ldscript.mips:272]: yes

File: sys/conf/options (hunk 29 above)
Comment (brooksdavis) [reply sys/conf/options:1022]: If we need it for files we can change it to opt_dontuse.h to prevent use.

Comment (jrtc27) [reply sys/conf/options:1022]: And for those cases should it not just be `CHERI_PURECAP`? The `_KERNEL` is redundant.

Comment (bsdjhb) [reply sys/conf/options:1022]: It is similar to CPU_CHERI in that they shouldn't be used in #ifdef's, only for sys/conf/files.  I'm not sure if we've actually cleaned up CPU_CHERI enough to move it to opt_dontuse.h, but CHERI_PURECAP_KERNEL should be clean enough to move.

I don't really care strongly what the name is, but it would be nice if CPU_CHERI and CHERI_PURECAP were a bit closer in names, perhaps CPU_CHERI should be CHERI_HYBRID instead?

Comment (jrtc27) [reply sys/conf/options:1022]: CPU_CHERI is _meant_ to be somewhat akin to CPU_QEMU_MALTA and CPU_BERI. It's used almost nowhere outside of sys/mips. But even in the current situation it's more like `__has_feature(capabilities)`, whereas CHERI_HYBRID would not be true for CHERI_PURECAP.

Comment (jrtc27) [reply sys/conf/options:1022]: i.e. really we need to split it out into a genuine `__has_feature(capabilities)` option for files and a CPU_CHERI to mean CHERI-MIPS (though maybe that can be dropped and just use CPU_BERI?).

Comment (bsdjhb) [reply sys/conf/options:1022]: Hmm, looking at sys/conf/files, CHERI_HYBRID doesn't work indeed.  CPU_CHERI really means either hybrid or purecap, it is the __has_feature() equivalent you mentioned.  CPU_CHERI is probably fine for that name-wise as CHERI itself is not BERI specific.  I guess I don't have a better name.  At least CHERI_PURECAP is shorter than `__CHERI_PURE_CAPABILITY__`.

File: sys/conf/ldscript.mips (hunk 27 above)
Comment (arichardson) [reply sys/conf/ldscript.mips:268]: I _think_ it should also be possible to use ALIGN(_BSS_ALIGN) and then pass `-Wl,-defsym=_BSS_ALIGN=0x10000` in the purecap case and `-Wl,-defsym=_BSS_ALIGN=8` otherwise.

Comment (arichardson) [reply sys/conf/ldscript.mips:268]: Just checked and this works with ld.lld.

**Review by brooksdavis — APPROVED** [review]
My issues have been addressed

**Comment by arichardson** [comment]
The 1MB padding doesn't matter for QEMU, and I don't think I'll be using a DE4 any time soon, so not using a variable for the alignment is fine for me.

File: sys/conf/ldscript.mips (hunk 23 above)
Comment (jrtc27) [reply sys/conf/ldscript.mips:54]: But why does this only happen for a purecap kernel? Hybrid kernel work just fine.

File: sys/conf/ldscript.mips [hunk 30]
```diff
@@ -176,6 +180,7 @@ SECTIONS
     }
   .rel.plt       : { *(.rel.plt)		}
   .rela.plt      : { *(.rela.plt)		}
+
```
Comment (jrtc27) [inline sys/conf/ldscript.mips:183]: Spurious diff.

File: sys/conf/options (hunk 29 above)
Comment (jrtc27) [reply sys/conf/options:1022]: This still needs moving to opt_dontuse.h.

File: sys/conf/kern.mk [hunk 31]
```diff
@@ -270,12 +270,15 @@ CFLAGS+=	-gdwarf-2
 #
 .if ${MACHINE_ABI:Mpurecap}
 
+.if defined(CHERI_SUBOBJECT_BOUNDS)
+CFLAGS+=	-Xclang -cheri-bounds=${CHERI_SUBOBJECT_BOUNDS}
 .if defined(CHERI_SUBOBJECT_BOUNDS_STATS)
```
Comment (jrtc27) [inline sys/conf/kern.mk:280]: Shouldn't this be non-nested? We still get CSetBounds for alloca's that contribute to stats IIRC.

File: sys/conf/kern.mk [hunk 32]
```diff
@@ -336,6 +339,7 @@ LD_EMULATION_i386=elf_i386_fbsd
 LD_EMULATION_mips= elf32btsmip_fbsd
 LD_EMULATION_mipshf= elf32btsmip_fbsd
 LD_EMULATION_mips64= elf64btsmip_fbsd
+LD_EMULATION_mips64c128= elf64btsmip_cheri_fbsd
```
Comment (jrtc27) [inline sys/conf/kern.mk:347]: Does it still need the `_cheri`? My LLD change should have made it unnecessary in general, not just RISC-V.

Comment (qwattash) [reply sys/conf/kern.mk:347]: Just checked, not needed anymore.

File: sys/conf/kern.mk (hunk 31 above)
Comment (bsdjhb) [reply sys/conf/kern.mk:282]: This just matches bsd.cheri.mk which was then copied into bsd.cpu.mk.  We should keep the subobject bounds bits consistent for userland and kernel I think.

**Comment by jrtc27** [comment]
@qwattash will you be filing a follow-up PR to deal with the unresolved comments?
